tests/storage/test_roommember.py

```diff
@@ -161,6 +161,60 @@ def test_get_joined_users_from_context(self):
         )
         self.assertEqual(users.keys(), {self.u_alice, self.u_bob})
 
+    def test_store_room_members_txn(self):
+        def _store_room_members_txn(txn, events, backfilled):
+            """Store a room member in the database."""
+
+            def non_null_str_or_none(val) -> str:
+                return val if isinstance(val, str) and "\u0000" not in val else None
+
+            self.store.db_pool.simple_insert_many_txn(
+                txn,
+                table="room_memberships",
+                values=[
+                    {
+                        "event_id": event.event_id,
+                        "user_id": "random_test_value",
+                        "sender": event.user_id,
+                        "room_id": event.room_id,
+                        "membership": event.membership,
+                        "display_name": non_null_str_or_none(
+                            event.content.get("displayname")
+                        ),
+                        "avatar_url": non_null_str_or_none(
+                            event.content.get("avatar_url")
+                        ),
+                    }
+                    for event in events
+                ],
+            )
+
+        room = self.helper.create_room_as(self.u_alice, tok=self.t_alice)
+        bob_event = self.get_success(
+            event_injection.inject_member_event(
+                self.hs, room, self.u_bob, Membership.JOIN
+            )
+        )
+
+        # Create an event with a null-containing string to pass to _store_room_members_txn
+        event, context = self.get_success(
+            event_injection.create_event(
+                self.hs,
+                room_id=room,
+                sender=self.u_alice,
+                prev_event_ids=[bob_event.event_id],
+                type="m.test.1",
+                content={"displayname": "bad\u0000null", "membership": Membership.JOIN},
+            )
+        )
```

Comment:
I usually prefer to avoid injecting events if you can avoid it (reserve that for when you need invalid events to be persisted for some reason, or to pretend you received one over federation).

Since you have alice's access token and everything, you can use something like

```
        new_profile = {"displayname": "ali\u0000ce"}
        self.helper.change_membership(
            room, self.u_alice, self.u_alice, "join", extra_data=new_profile, tok=self.t_alice
        )
```

Reply:
The call to `change_membership` does indeed break without the change I made to `_store_room_members_txn` and passes with the change, so hopefully this is sufficient? Thanks for pointing out the use cases for injecting events, it's helpful to be told these sorts of things :)

Reply:
Maybe it would be worth also querying the state of the database and checking that they contain Nones instead of something else.
(It just feels weird to have an assert-less test in some strange way, but it's also worth noting that the behaviour is replacing the name with None rather than e.g. leaving it as it was before, if that makes sense?)

Reply:
Totally makes sense, I shall give that a go.